### ISSUE TITLE
<type_traits>: `is_constructible` reports wrong value for types with conversion operator to immovable types

### ISSUE THREAD

**Issue by Chlorie** [microsoft/STL#2672]
Code sample:
(Godbolt link: https://godbolt.org/z/j7TK5sfcr)

``` cpp
#include <type_traits>

struct constructor {};

struct immovable
{
    immovable() = default;
    immovable(constructor) {}
    immovable(const immovable&) = delete;
};

struct converter
{
    operator immovable() { return {}; }
};

int main()
{
    static_assert(std::is_constructible_v<immovable, constructor>);
    static_assert(std::is_constructible_v<immovable, converter>);
}
```

Both gcc and clang accept this but MSVC fires the second `static_assert`. This hinders emplacing immovable types with copy elision into a `variant` or algebraic types like that.

**Comment by CaseyCarter** [comment]
> Thanks, I created [DevCom-10024578](https://developercommunity.visualstudio.com/t/10024578) and [DevCom-10024617](https://developercommunity.visualstudio.com/t/10024617)

Thanks for jumping on this @fsb4000, I agree this looks like a bug in the compiler intrinsic that we use to implement `is_constructible`.

**Comment by frederick-vs-ja** [comment]
I (again!?) guess a workaround is available (like the following), but not sure whether it's welcome.

```C++
template <class _Void, class _Ty, class... _Args>
_INLINE_VAR constexpr bool _Is_class_constructible_impl = false;

template <class _Ty, class... _Args>
_INLINE_VAR constexpr bool _Is_class_constructible_impl<std::void_t<decltype(_Ty(std::declval<_Args>()...))>, T, _Args...>
    = true;
    
template <class _Ty, class... _Args>
_INLINE_VAR constexpr bool is_constructible_v =
    __is_class(_Ty) ? is_class_constructible_impl_v<void, _Ty, _Args...> : __is_constructible(_Ty, _Args...);
```
----

BTW, MSVC also misprocessed `std::is_constructible_v<converter, immovable>`. I failed to fiidn a workaround for this due to another (?) bug ([Compiler Explorer](https://godbolt.org/z/PrMjPcTj6)).

```C++
struct constructor {};

struct immovable
{
    immovable() = default;
    immovable(constructor) {}
    immovable(const immovable&) = delete;
};

struct converter
{
    operator immovable() { return {}; }
};

int main()
{
    // Constructibility
    immovable xx [[maybe_unused]](constructor{});
    static_cast<immovable>(constructor{});
    immovable(constructor{});

    immovable yy [[maybe_unused]](converter{});
    static_cast<immovable>(converter{});
    immovable(converter{});

    // Convertibility
    immovable zz [[maybe_unused]] = constructor{};
    [](immovable){}(constructor{});
    []() -> immovable { return constructor{}; };

    immovable ww [[maybe_unused]] = converter{};
    [](immovable){}(converter{}); // MSVC is buggy here
    []() -> immovable { return converter{}; };
}
```

**Comment by fsb4000** [comment]
> I failed to find a workaround for this due to another (?) bug

I decided not create another issue and I added your example to the current DevCom issues.

**Comment by fsb4000** [comment]
DevCom-10024578 : "Fixed - Pending Release"

**Comment by fsb4000** [comment]
DevCom-10024617 also : "Fixed - Pending Release"

**Comment by frederick-vs-ja** [comment]
This issue seems able to be closed now.
- original issue is fixed in 19.33 ([Godbolt link](https://godbolt.org/z/7aEfax5e7));
- the other is fixed in 19.39 ([Godbolt link](https://godbolt.org/z/djaW3jeo3)).
